docs/features/analysis.md

```diff
@@ -873,3 +873,43 @@ For success conditions that need to evaluate a numeric return value the `asInt`
             value: "Bearer {{ args.api-token }}"
         jsonPath: "{$.results.successPercent}" 
 ```
+
+
+## Datadog Metrics
+
+A [Datadog](https://www.datadoghq.com/) query can be used to obtain measurements for analysis.
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: AnalysisTemplate
+metadata:
+  name: loq-error-rate
+spec:
+  args:
+  - name: service-name
+  metrics:
+  - name: error-rate
+    interval: 5m
+    successCondition: result <= 0.01
+    failureLimit: 3
+    provider:
+      datadog:
+        address: https://api.datadoghq.com
+        interval: 5m
+        query: |
+          sum:requests.error.count{service:{{args.service-name}}} /
+          sum:requests.request.count{service:{{args.service-name}}}
+```
+
+Datadog api and app tokens can be configured in a kubernetes secret in argo-rollouts namespace.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: wavefront-api-tokens
+type: Opaque
+data:
+  api-key: <datadog-api-key>
+  app-key: <datadog-app-key>
```

Comment:
1. docs still say "wavefront-api-tokens". I think this should be `datadog-auth-keys`
2. don't we need to allow multiple API/APP keys to be configured?

Reply:
(1) - oops! It should be `datadog`. Fixed now

(2) - I don't thinks so. All key pairs are assigned to the organisation, and using them you can pull any metrics from your organisation. Our organisation encourages teams to own their own keys, but team a can use its keys to access metrics reported by team b. We could make the secret name configurable, and you specify it as an AnalysisTemplate parameter, but I don't see the value in doing that.

Reply:
OK I'll take your word for it, but I'll also pose the question in #argo-rollouts slack in case there's users who may need to configure multiple, in case there's something we're missing.

Reply:
Just to be clear:

With the proposed spec, we allow the DataDog `address` to be specified in the metric (so it can vary between AnalysisTemplates), but API/APP keys are configured centrally in the controller. This presumes that the same API/APP key can be used for different DataDog URLs. This may very well be possible with DataDog (I don't know the answer to this), but I wanted to call that out as a potential flaw.

Reply:
That's a good point. No API/APP keys are not cross-site. In fact I believe an organisation is in either the EU or US site. Although not a secret, for ease of configuration we could move these to the `datadog` secret (alongside `app_key` and `api_key`). What do you think?

Reply:
Yes. I agree we should make the address and api/app keys centrally configured, and remove the `address` field from the metric. I'm thinking the Secret should look something like:

```yaml
data:
  default: |
    address: https://api.datadoghq.com
    api-key: igq7imbxlm97vnlw61bo
    app-key: 35bu8ofvc0osyjmqxtgi
```

This also leaves the door open so if in the future, multiple addresses and/or app keys could to be specified in the same cluster. e.g.:

```yaml
data:
  us-app1: |
    address: https://api.datadoghq.com
    api-key: igq7imbxlm97vnlw61bo
    app-key: 35bu8ofvc0osyjmqxtgi
  us-app2: |
    address: https://api.datadoghq.com
    api-key: igq7imbxlm97vnlw61bo
    app-key: cb7ehxo1jwagyceth8v8
  eu: |
    address: https://api.datadoghq.eu
    api-key: cb7ehxo1jwagyceth8v8
    app-key: timk0do15pafuvmfwimg
```

And the names `us-app1` and `us-app2` could be referenced in the metric.

But for MVP, for simplicity, we can go out with a single configured DataDog region and single api/app key combination.

Reply:
I've implemented this, though I'm starting to think perhaps we should have just one level (e.g. a secret per app/region), and then reference the name of the secret in the metric.

```
apiVersion: v1
kind: Secret
metadata:
   name: datadog
data:
  address: https://api.datadoghq.com
  api-key: igq7imbxlm97vnlw61bo
  app-key: 35bu8ofvc0osyjmqxtgi
---
apiVersion: v1
kind: Secret
metadata:
   name: datadog-app2
data:
  address: https://api.datadoghq.eu
  api-key: cb7ehxo1jwagyceth8v8
  app-key: timk0do15pafuvmfwimg
```

Not too fussed though; I think in most use cases organisations will only have one region, and will only have key pair used by rollouts.

Reply:
> though I'm starting to think perhaps we should have just one level (e.g. a secret per app/region), and then reference the name of the secret in the metric.

Actually I agree with you. A secret per app/region sounds better as it doesn't put a special datastructure in the secret payload. Then things like kustomize patching the secret will just work. Let's go with your suggestion, but we don't have to do anything yet with regards to referencing until someone has a need to reference multiple regions.